lp-lsp/lp_lsp.ml

```diff
@@ -175,16 +187,103 @@ let get_goals ~doc ~line ~pos =
         LIO.log_error "get_goals" ("call: "^ls);
         res
       ) doc.Lp_doc.nodes in
-  Option.map
-    (fun node -> Format.asprintf "%a" Proof.pp_goals node.goals) node
-
-let do_hover ofmt ~id params =
+  let goalsList = match node with
+    | None -> []
+    | Some n -> n.goals in
+  let goals = match (List.find_opt (fun (_, loc) -> in_range ?loc (line,pos)) goalsList) with
+    | None -> None
+    | Some (v,_) -> Some v in
+  goals
+
+let do_goals ofmt ~id params =
   let uri, line, pos = get_docTextPosition params in
   let doc = Hashtbl.find completed_table uri in
-  get_goals ~doc ~line ~pos |> Option.iter (fun goals ->
-      let result = `Assoc [ "contents", `String goals] in
-      let msg = LSP.mk_reply ~id ~result in
-      LIO.send_json ofmt msg)
+  let goals = get_goals ~doc ~line ~pos in
+  let result = LSP.json_of_goals goals in
+  let msg = LSP.mk_reply ~id ~result in
+  LIO.send_json ofmt msg
+
+let myfail (msg : string) =
+  LIO.log_error "fail" msg;
+  failwith msg
+
+let get_token tokens pos =
```

Comment:
Some comment explaining what this function does would be welcome.

Reply:
I've done what I could, the function should likely be refactored in full.